src/rgw/driver/rados/rgw_trim_bilog.cc

```diff
@@ -1443,3 +1618,62 @@ int bilog_trim(const DoutPrefixProvider* p, rgw::sal::RadosStore* store,
   }
   return r;
 }
+
+int list_deleted_buckets(const DoutPrefixProvider* dpp,
+                        rgw::sal::RadosStore* store,
+                        std::set<std::string>& buckets,
+                        optional_yield y)
+{
+  auto& pool = store->svc()->zone->get_zone_params().log_pool;
+  librados::Rados& rados = *store->getRados()->get_rados_handle();
+  rgw_rados_ref ref;
+  int r = rgw_get_rados_ref(dpp, &rados, rgw_raw_obj(pool, oid_prefix), &ref);
+  if (r < 0) {
+    return r;
+  }
+
+  std::set<std::string> keys;
+  std::string start_after;
+  constexpr int max_items = 1024;
+  bool more = true;
+  int ret_val = 0;
+  while (more) {
+    librados::ObjectReadOperation op;
+    op.omap_get_keys2(start_after, max_items, &keys, &more, &ret_val);
+    r = rgw_rados_operate(dpp, ref.ioctx, ref.obj.oid, &op, nullptr, y);
+    if (r < 0) {
+      return r;
+    }
+    if (ret_val < 0) {
+      return ret_val;
+    }
+
+    buckets.merge(std::move(keys));
+  }
+  return 0;
+}
+
+int remove_deleted_buckets(const DoutPrefixProvider *dpp, 
+                          rgw::sal::RadosStore* store,
+                          const std::set<std::string> buckets,
+                          optional_yield y)
+{
+  auto& pool = store->svc()->zone->get_zone_params().log_pool;
+  librados::Rados& rados = *store->getRados()->get_rados_handle();
+  rgw_rados_ref ref;
+  int r = rgw_get_rados_ref(dpp, &rados, rgw_raw_obj(pool, oid_prefix), &ref);
+  if (r < 0) {
+    return r;
+  }
+
+  librados::ObjectWriteOperation op;
+  for(const auto& key : buckets) {
+    op.omap_rm_keys(std::set<std::string>({key}));
+    r = rgw_rados_operate(dpp, ref.ioctx, ref.obj.oid, &op, y);
```

Comment:
remove_deleted_buckets() is only removing these omap keys, but the bucket instance and index still exist, presumably with objects remaining. how do you see picture the admin cleaning those up? can we try to document this process?

Reply:
right, this option `--remove-deleted` has to be used only once the admin has deleted the objects inside the bucket and after the trim has finished processing the bucket. but probably its hard for admin to say if the trim has finished processing the bucket. since we didn't want this to be a part of trimming process, I added it as a cli option. maybe sufficient documentation around this option is enough?

Reply:
> maybe sufficient documentation around this option is enough?

could be. or maybe the `radosgw-admin bucket rm` command could try to remove the bucket from this omap object automatically

Reply:
> > maybe sufficient documentation around this option is enough?
> 
> could be. or maybe the `radosgw-admin bucket rm` command could try to remove the bucket from this omap object automatically

the bucket omap entry is added only after trim sees `BucketLogType::Deleted` following a delete bucket execution. when do you suggest   `radosgw-admin bucket rm` has to be run to remove the entry? i could, instead, write up a separate command to handle its removal.

Reply:
if we're documenting the use of `radosgw-admin bucket rm --bucket {name} --bucket-id {instance-id} --purge-objects` to clean these up, it would be nice if that automatically unlinked the bucket from this deleted_buckets object

Reply:
what if an admin has used awscli or something else to delete the bucket? do we want them to run` bucket rm ` for the omap entry cleanup?

Reply:
to calrify, i'm talking about deleted buckets that trim discovers to still contain objects. these get linked into `deleted_buckets`, so become visible to the `radosgw-admin bucket list-deleted` command

in response, the admin should check whether this bucket contains any important objects that need to be backed up elsewhere before completing the bucket's cleanup. once they're confident that they're recovered everything that was important, they can complete the cleanup with `radosgw-admin bucket rm --bucket {name} --bucket-id {instance-id} --purge-objects`. the bucket entrypoint was already deleted, so they have to specify the --bucket-id manually to find it. that will delete the remaining objects, then delete the bucket index object and instance metadata

after doing that for every bucket, they could run `bucket list-deleted --remove-deleted` to clear the list. but there's a risk that trim added another bucket to the list since they last ran `bucket list-deleted`, and `--remove-deleted` would clear that before they have a chance to do the cleanup

so i'm suggesting that the `bucket rm` command, on success, unlinks the bucket from this `deleted_objects` list to avoid these races and simplify the process for admins

Reply:
oh ok I don't think we spoke about using `bucket rm ` earlier to clean up stuff locally. I was working with the assumption that once an admin moves/removes the objects inside the bucket, the next trim run will take care of the instance cleanup. this makes sense.

Reply:
done. need to work on documentation.

Reply:
updated PendingReleaseNotes. will run it through teuthology. do you have any more review comments @cbodley